### PR TITLE
Suggestions for assembler config code [skip ci]

### PR DESCRIPTION
**Pull request by standage**
My suggestions for the config code were easier to submit as a PR than comments. See below. (Note: this is a PR into the configfile branch, not the main branch.)

### CODE REVIEW

File: yeat/config.py [hunk 1]
```diff
@@ -11,59 +11,44 @@
...
+            message = "Duplicate assembly configuration: please check config file"
+            raise AssemblyConfigurationError(message)
+        configlist = [AssemblerConfig.from_json(entry) for entry in configdata]
+        return configlist
 
     @staticmethod
-    def check_algorithm(algorithm, algorithms):
-        if algorithm not in ASSEMBLY_ALGORITHMS:
-            message = f"Found unsupported assembly algorithm in config settings: [[{algorithm}]]!"
-            raise ValueError(message)
```
Comment (standage) [inline yeat/config.py:47]: More appropriately done in the constructor.

File: yeat/config.py [hunk 2]
```diff
@@ -11,59 +11,44 @@
...
+class AssemblyConfigurationError(ValueError):
     pass
 
 
 class AssemblerConfig:
-    def __init__(self, assembler, extra_args=""):
-        self.assembler = assembler
+    def __init__(self, algorithm, extra_args=""):
+        if algorithm not in ALGORITHMS:
+            raise AssemblyConfigurationError(f"Unsupported assembly algorithm '{algorithm}'")
```
Comment (standage) [inline yeat/config.py:24]: Here.

**Comment by standage** [comment]
Comments pending.

File: yeat/config.py [hunk 3]
```diff
@@ -11,59 +11,44 @@
...
+        return configlist
 
     @staticmethod
-    def check_algorithm(algorithm, algorithms):
-        if algorithm not in ASSEMBLY_ALGORITHMS:
-            message = f"Found unsupported assembly algorithm in config settings: [[{algorithm}]]!"
-            raise ValueError(message)
-        if algorithm in algorithms:
-            message = f"Found duplicate assembly algorithm in config settings: [[{algorithm}]]!"
-            raise ValueError(message)
```
Comment (standage) [inline yeat/config.py:50]: This function isn't clearly factored. Understanding the code here requires the reader to also know what's going on in the calling function (unnecessarily tight coupling). In any case, if we move the "supported algorithm" check to the constructor, the "duplicate" check is easily moved into the loading function and made simpler with set operations.

File: yeat/config.py [hunk 4]
```diff
@@ -11,59 +11,44 @@
...
-            algorithm = d["assembler"]
-            AssemblerConfig.check_algorithm(algorithm, algorithms)
-            algorithms.add(algorithm)
-            assemblers.append(AssemblerConfig.from_json(d))
-        return assemblers
+        configdata = json.load(instream)
+        for entry in configdata:
+            AssemblerConfig.validate(entry)
+        algorithms = [entry["algorithm"] for entry in configdata]
+        if len(algorithms) > len(set(algorithms)):
```
Comment (standage) [inline yeat/config.py:38]: Duplicate check on this line.

File: yeat/config.py [hunk 5]
```diff
@@ -11,59 +11,44 @@
...
-        extrakeys = set()
-        for key in expectedkeys:
-            if key not in config:
-                missingkeys.add(key)
-        for key in config:
-            if key not in expectedkeys:
-                extrakeys.add(key)
+    def validate(config):
+        missingkeys = set(["algorithm", "extra_args"]) - set(config.keys())
+        extrakeys = set(config.keys()) - set(["algorithm", "extra_args"])
```
Comment (standage) [inline yeat/config.py:47]: Set operations also make this code much less verbose.

Comment (danejo3) [reply yeat/config.py:47]: This was a great idea!

**Pull request merged**